matrix-e2e.c

```diff
@@ -0,0 +1,1767 @@
+        MatrixHashKeyOlm *key = g_new0(MatrixHashKeyOlm, 1);
+        key->sender_key = g_strdup(sender_key);
+        key->sender_id = g_strdup(sender_id);
+        /* We loaded entries where there were none before, set the hash */
+        g_hash_table_insert(conn->e2e->olm_session_hash, key, cur_entry);
+    }
+
+    return cur_entry;
+
+err:
+    g_free(pickle);
+    free_matrix_olm_session(cur_entry, FALSE);
+    return NULL;
+}
+
+/* Update an existing OLM session in the database */
+static int update_olm_session(MatrixConnectionData *conn,
+                              MatrixOlmSession *mos)
+{
+    size_t pickle_len = olm_pickle_session_length(mos->session);
+    gchar *pickle = g_malloc(pickle_len+1);
+    sqlite3_stmt *dbstmt = NULL;
+    int ret = -1;
+
+    pickle_len = olm_pickle_session(mos->session, "!", 1, pickle, pickle_len);
+    if (pickle_len == olm_error()) {
+        purple_debug_warning("matrixprpl",
+                "%s: Failed to pickle session for %s/%s: %s\n",
+                __func__, mos->sender_id, mos->sender_key,
+                olm_session_last_error(mos->session));
+        goto err;
+    }
+    pickle[pickle_len] = '\0';
+
+    const char *query ="UPDATE olmsessions SET session_pickle=? "
+                       "WHERE sender_name = ? AND sender_key = ? AND "
+                       "ROWID = ?";
+    ret = sqlite3_prepare_v2(conn->e2e->db, query, -1, &dbstmt, NULL);
+    if (ret != SQLITE_OK || !dbstmt) {
+        purple_debug_warning("matrixprpl",
+                "%s: Failed to prep update %d '%s'\n",
+                __func__, ret, query);
+        ret = -1;
+        goto err;
+    }
+    ret = sqlite3_bind_text(dbstmt, 1, pickle, -1, NULL);
+    if (ret == SQLITE_OK) {
+        ret = sqlite3_bind_text(dbstmt, 2, mos->sender_id, -1, NULL);
+    }
+    if (ret == SQLITE_OK) {
+        ret = sqlite3_bind_text(dbstmt, 3, mos->sender_key, -1, NULL);
+    }
+    if (ret == SQLITE_OK) {
+        ret = sqlite3_bind_int64(dbstmt, 4, mos->unique);
+    }
+
+    if (ret != SQLITE_OK) {
+        purple_debug_warning("matrixprpl",
+                             "%s: Failed to bind %d\n", __func__, ret);
+        ret = -1;
+        goto err;
+    }
+
+    ret = sqlite3_step(dbstmt);
+    if (ret != SQLITE_DONE) {
+        purple_debug_warning("matrixprpl",
+                             "%s: Update failed %d (%s)\n", __func__,
+                ret, query);
+        ret = -1;
+        goto err;
+
+    }
+    ret = 0;
+
+err:
+    sqlite3_finalize(dbstmt);
+    g_free(pickle);
+
+    return ret;
+}
+
+static void key_upload_callback(MatrixConnectionData *conn,
+                                gpointer user_data,
+                                struct _JsonNode *json_root,
+                                const char *body,
+                                size_t body_len, const char *content_type);
+
+/* Really clear an area of memory */
+static void clear_mem(volatile char *data, size_t len)
```

Comment:
I think this function should set random bytes, not set to zeroes.

Reply:
I don't think there's any benefit to using random bytes compared to zeroes; it's not like overwriting magic on a hard drive.

Reply:
Well to an attacker, receiving zeroes is a sure sign that they have not found a real key, its basically obfuscation at its finest :)